sql/parser/eval.go

```diff
@@ -259,6 +263,151 @@ func (d DString) String() string {
 	return StrVal(d).String()
 }
 
+// DDate is the date Datum.
+type DDate struct {
+	time.Time
+}
+
+// Type implements the Datum interface.
+func (d DDate) Type() string {
+	return "date"
+}
+
+// Compare implements the Datum interface.
+func (d DDate) Compare(other Datum) int {
+	if other == DNull {
+		// NULL is less than any non-NULL value.
+		return 1
+	}
+	v, ok := other.(DDate)
+	if !ok {
+		panic(fmt.Sprintf("unsupported comparison: %s to %s", d.Type(), other.Type()))
+	}
+	if d.Before(v.Time) {
+		return -1
+	}
+	if v.Before(d.Time) {
+		return 1
+	}
+	return 0
+}
+
+// Next implements the Datum interface.
+func (d DDate) Next() Datum {
+	return DDate{Time: d.Time.AddDate(0, 0, 1)}
+}
+
+// IsMax implements the Datum interface.
+func (d DDate) IsMax() bool {
+	// Adding a day overflows to a smaller value.
+	return d.Time.After(d.Time.AddDate(0, 0, 1))
```

Comment:
If we had a Prev() function it would make sense to use Prev() and Next() in these methods. Anyway I've changes IsMax to use Next and left IsMin as is.